### ISSUE TITLE
Running out of memory when training on Metacentrum

### ISSUE THREAD

**Issue by mtichavsky** [simon-smida/KNN-project#21]
Happend in 2nd epoch of training with minibatch of 16 (had other run when it didn't happen). When running with minibatch 32, happend towards the end of 1st epoch IIRC.

Logs:

```
Iteration 15800, average loss: 0.7723666673898697, prediction accuracy 0.0
Used GPU memory 0.12313270568847656 GB
Iteration 15850, average loss: 0.688846404850483, prediction accuracy 0.00125
Used GPU memory 0.12084293365478516 GB
Iteration 15900, average loss: 0.650011313855648, prediction accuracy 0.00125
Used GPU memory 0.12572288513183594 GB
Iteration 15950, average loss: 0.6678449141979218, prediction accuracy 0.00125
Used GPU memory 0.1299142837524414 GB
Iteration 16000, average loss: 0.7228622630238533, prediction accuracy 0.0
Used GPU memory 0.12195682525634766 GB
Iteration 16050, average loss: 0.7352228716015816, prediction accuracy 0.0
Used GPU memory 0.12009143829345703 GB
Iteration 16100, average loss: 0.6400340580940247, prediction accuracy 0.00125
Used GPU memory 0.1187143325805664 GB
Iteration 16150, average loss: 0.6911307755112648, prediction accuracy 0.0
Used GPU memory 0.11684513092041016 GB
Iteration 16200, average loss: 0.6559812724590302, prediction accuracy 0.00125
Used GPU memory 0.11876296997070312 GB
Iteration 16250, average loss: 0.7483058626949787, prediction accuracy 0.0
Used GPU memory 0.12333393096923828 GB
Iteration 16300, average loss: 0.657174420952797, prediction accuracy 0.0025
Used GPU memory 0.12521839141845703 GB
Iteration 16350, average loss: 0.6757675641775132, prediction accuracy 0.0025
Used GPU memory 0.12101078033447266 GB
Traceback (most recent call last):
  File "/scratch.ssd/tichavskym/job_1260483.pbs-m1.metacentrum.cz/knn/train_ecapa.py", line 96, in <module>
    outputs = model(x)
  File "/scratch.ssd/tichavskym/job_1260483.pbs-m1.metacentrum.cz/knn/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/scratch.ssd/tichavskym/job_1260483.pbs-m1.metacentrum.cz/knn/models/ecapa.py", line 534, in forward
    x = self.asp(x, lengths=lengths)
  File "/scratch.ssd/tichavskym/job_1260483.pbs-m1.metacentrum.cz/knn/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/scratch.ssd/tichavskym/job_1260483.pbs-m1.metacentrum.cz/knn/models/ecapa.py", line 293, in forward
    mean, std = _compute_statistics(x, attn)
  File "/scratch.ssd/tichavskym/job_1260483.pbs-m1.metacentrum.cz/knn/models/ecapa.py", line 262, in _compute_statistics
    (m * (x - mean.unsqueeze(dim)).pow(2)).sum(dim).clamp(eps)
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 1.33 GiB (GPU 0; 44.35 GiB total capacity; 36.61 GiB already allocated; 1.30 GiB free; 42.70 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
make: *** [Makefile:18: train] Error 1
Saving the results into the storage directory /storage/brno12-cerit/home/tichavskym/results/train-2024-04-30-19-24
/bin/bash: /cvmfs/software.metacentrum.cz/spack1/software/ncurses/linux-debian10-x86_64/6.2-intel-hucms2/lib/libtinfo.so.6: no version information available (required by /bin/bash)
```

Also, I should fix clearing the iteration count.

Note there's some older pytorch version on Metacentrum.

**Comment by mtichavsky** [comment]
```
Memory allocated: 0.10462427139282227 GB, memory reserved: 0.529296875 GB
Memory allocated: 0.09884500503540039 GB, memory reserved: 0.51953125 GB
Memory allocated: 0.10121011734008789 GB, memory reserved: 0.50390625 GB
Memory allocated: 0.09503030776977539 GB, memory reserved: 0.69140625 GB
Memory allocated: 0.09777688980102539 GB, memory reserved: 6.25390625 GB
Memory allocated: 0.10021829605102539 GB, memory reserved: 0.49609375 GB
Memory allocated: 0.09405755996704102 GB, memory reserved: 4.3515625 GB
Memory allocated: 0.09497976303100586 GB, memory reserved: 0.451171875 GB
Memory allocated: 0.09714746475219727 GB, memory reserved: 0.802734375 GB
Memory allocated: 0.09497976303100586 GB, memory reserved: 4.341796875 GB
Memory allocated: 0.09544992446899414 GB, memory reserved: 4.853515625 GB
Memory allocated: 0.09554529190063477 GB, memory reserved: 0.5625 GB
Memory allocated: 0.10012292861938477 GB, memory reserved: 0.49609375 GB
Memory allocated: 0.09613656997680664 GB, memory reserved: 14.009765625 GB
Memory allocated: 0.09497976303100586 GB, memory reserved: 7.837890625 GB
Memory allocated: 0.10334634780883789 GB, memory reserved: 0.515625 GB
Memory allocated: 0.09497976303100586 GB, memory reserved: 0.439453125 GB
Memory allocated: 0.09929323196411133 GB, memory reserved: 9.427734375 GB
Memory allocated: 0.09527826309204102 GB, memory reserved: 4.701171875 GB
Memory allocated: 0.10328912734985352 GB, memory reserved: 11.6640625 GB
Memory allocated: 0.09995126724243164 GB, memory reserved: 0.494140625 GB
Memory allocated: 0.09810113906860352 GB, memory reserved: 11.658203125 GB
Memory allocated: 0.09836816787719727 GB, memory reserved: 7.568359375 GB
Memory allocated: 0.09382867813110352 GB, memory reserved: 7.73046875 GB
Memory allocated: 0.09497976303100586 GB, memory reserved: 13.146484375 GB
Memory allocated: 0.10620737075805664 GB, memory reserved: 0.66796875 GB
Memory allocated: 0.09497976303100586 GB, memory reserved: 7.791015625 GB
Memory allocated: 0.10176324844360352 GB, memory reserved: 11.6171875 GB
Memory allocated: 0.09867334365844727 GB, memory reserved: 8.291015625 GB
Memory allocated: 0.09554529190063477 GB, memory reserved: 9.509765625 GB
...
Memory allocated: 0.11168146133422852 GB, memory reserved: 0.755859375 GB
Memory allocated: 0.09717702865600586 GB, memory reserved: 14.431640625 GB
Memory allocated: 0.09497976303100586 GB, memory reserved: 14.431640625 GB
Memory allocated: 0.10430002212524414 GB, memory reserved: 18.021484375 GB
Memory allocated: 0.09539270401000977 GB, memory reserved: 13.966796875 GB
Memory allocated: 0.10185861587524414 GB, memory reserved: 12.228515625 GB
Memory allocated: 0.09867334365844727 GB, memory reserved: 17.515625 GB
Memory allocated: 0.10138177871704102 GB, memory reserved: 12.13671875 GB
Memory allocated: 0.10132455825805664 GB, memory reserved: 19.83203125 GB
Memory allocated: 0.09760522842407227 GB, memory reserved: 19.83203125 GB
Memory allocated: 0.09770059585571289 GB, memory reserved: 14.6875 GB

Epoch 2:
Memory allocated: 0.09789133071899414 GB, memory reserved: 19.283203125 GB
Memory allocated: 0.09821557998657227 GB, memory reserved: 21.203125 GB
Memory allocated: 0.09802484512329102 GB, memory reserved: 29.16015625 GB
Memory allocated: 0.09778642654418945 GB, memory reserved: 29.16015625 GB
Memory allocated: 0.09717702865600586 GB, memory reserved: 29.16015625 GB
...
Memory allocated: 0.09955072402954102 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.10616922378540039 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.10791921615600586 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.09438180923461914 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.10218286514282227 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.09434366226196289 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.09827280044555664 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.10679864883422852 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.09602212905883789 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.10311746597290039 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.09609842300415039 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.11145257949829102 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.09501123428344727 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.09543085098266602 GB, memory reserved: 29.91015625 GB
Memory allocated: 0.09876871109008789 GB, memory reserved: 12.068359375 GB
Memory allocated: 0.10225915908813477 GB, memory reserved: 28.79296875 GB
Memory allocated: 0.09569787979125977 GB, memory reserved: 34.294921875 GB
Memory allocated: 0.09717702865600586 GB, memory reserved: 34.294921875 GB
Memory allocated: 0.09359979629516602 GB, memory reserved: 34.294921875 GB
Memory allocated: 0.09717702865600586 GB, memory reserved: 34.294921875 GB
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 1.33 GiB (GPU 0; 44.34 GiB total capacity; 36.61 GiB already allocated; 1.23 GiB free; 42.76 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

**Comment by mtichavsky** [comment]
> It seems that “reserved in total” is memory “already allocated” to tensors + memory cached by PyTorch. When a new block of memory is requested by PyTorch, it will check if there is sufficient memory left in the pool of memory which is not currently utilized by PyTorch (i.e. total gpu memory - “reserved in total”). If it doesn’t have enough memory the allocator will try to clear the cache and return it to the GPU which will lead to a reduction in “reserved in total”, however it will only be able to clear blocks on memory in the cache of which no part is currently allocated. If any of the block is allocated to a tensor it won’t be able to return it to GPU. Thus you can have scenarios where the tensor allocated memory + the gpu free memory is much less than the gpu total memory, because the cache is holding some which it cannot release

That's why cuda.empty_cache doesn't work probably.
Comment found [in this thread](https://discuss.pytorch.org/t/how-does-reserved-in-total-by-pytorch-work/70172/8) and references [this issue.](https://github.com/pytorch/pytorch/issues/35901)

From the issue:

>  I have read that using variable length inputs can increase the chances of fragmentation but here the input is of constant length.

Definitely our case rn.

> I can solve this problem completely by calling torch.cuda.empty_cache() directly after the large tensor is deleted (thus not leaving a large segment hanging around which becomes fragmented), but it would be nice if their was a better solution.

Maybe few of those recordings are very long which causes our troubles similar to theirs. How expensive would be calling empty cache more often?

**Comment by mtichavsky** [comment]
Suggested solutions:

The quick one:

- [x] launch train script for each iteration separately -> **still failing though**

The better one:

- [ ] max_split_size_mb
- [ ] Use the latest pytorch version on metacentrum, see [docs](https://docs.metacentrum.cz/software/sw-list/pytorch/)
- [x] Look cut the longest recordings into shorter ones (first gather some statistics about their sizes)

**Comment by mtichavsky** [comment]
Fixed by #28
